File: 3. BinarySearch/MissingAndRepeatedNumber/MissingAndRepeatedNum.java
class MissingAndRepeatedNum{
    public static void main(String[] args) {
        int arr[] = new int[]{2, 3, 2, 1, 5};
        findRepeatedMissingNum(arr);
    }
    public static void findRepeatedMissingNum(int[] arr){

        // TC: O(n^2)
        // SC: O(1)
 
		int repeating = -1;
		int missing = -1;
	    for (int i = 1; i <= arr.length; i++ ) {
		    int count = 0;
		    for (int j = 0; j < arr.length; j++) {
			    if (i == arr[j]) {
				    count++;
                }
			}
            if (count == 2) {
                repeating = i;
            }
            if (count == 0) {
                missing = i;
            }
		}
        System.out.println("Missing: "+missing);
        System.out.println("Repeating: "+repeating);
        
    }
}

File: 3. BinarySearch/MissingAndRepeatedNumber/MissingAndRepeatedNum.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{2, 3, 2, 1, 5}
	missing, repeating := findNumber(arr)
	fmt.Println("Missing: ", missing, " & ", "Repeating: ", repeating)
}

var findNumber = func(arr []int) (int, int) {

	// TC: O(n)
	// SC: O(n)
	var (
		repeating int = -1
		missing   int = -1
	)
	m := map[int]int{}
	for i := 1; i <= len(arr); i++ {
		m[i] = 0
	}
	fmt.Println(m)
	for i := 0; i < len(arr); i++ {
		m[arr[i]]++
	}
	for key := range m {
		if m[key] == 2 {
			repeating = key
		}
		if m[key] == 0 {
			missing = key
		}
	}
	fmt.Println(m)
	return missing, repeating
}

func binarySearch(arr []int, element int) (int, bool) {
	var low int = 0
	var high int = len(arr) - 1
	var mid int
	for low <= high {
		var mid int = (low + high) / 2
		if mid == element {
			return mid, true
		} else if arr[mid] > element {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return mid, false
}
